language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java

```diff
@@ -78,31 +80,33 @@ public static class Decoders {
           return value;
         };
 
-    public static final JsonLfDecoder<BigDecimal> decimal =
-        r -> {
-          r.expectIsAt(
-              "decimal",
-              JsonToken.VALUE_NUMBER_INT,
-              JsonToken.VALUE_NUMBER_FLOAT,
-              JsonToken.VALUE_STRING);
-          BigDecimal value = null;
-          try {
-            value = new BigDecimal(r.parser.getText());
-          } catch (NumberFormatException e) {
-            r.parseExpected("decimal", e);
-          } catch (IOException e) {
-            r.parseExpected("decimal", e);
-          }
-          r.moveNext();
-          return value;
-        };
+    public static JsonLfDecoder<BigDecimal> decimal(int scale) {
+      return r -> {
+        r.expectIsAt(
+            "decimal",
+            JsonToken.VALUE_NUMBER_INT,
+            JsonToken.VALUE_NUMBER_FLOAT,
+            JsonToken.VALUE_STRING);
+        BigDecimal value = null;
+        try {
+          value = new BigDecimal(r.parser.getText());
+          if (value.scale() > scale) value = value.setScale(scale, RoundingMode.HALF_EVEN);
```

Comment:
if I compare that to the JSON API implementation it seems to do this check: `x.precision - x.scale <= maxPrecision - scale`. Why is your check silghtly different?

Reply:
IIUC that's because this condition
https://github.com/digital-asset/daml/blob/37ffc7dd1654ff8b6c19f35b8b98e43270113799/daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/NumericModule.scala#L269
is checking for whether the value is within bounds, and if so, the `setScale` (i.e. rounding) is done unconditionally. Here I don't actually have a bounds check, but I only round if needed.